processing/src/main/java/org/apache/druid/segment/vector/VectorSelectorUtils.java

```diff
@@ -47,14 +47,38 @@
       retVal = new boolean[offset.getMaxVectorSize()];
     }
 
-    // Probably not super efficient to call "get" so much, but, no worse than the non-vectorized version.
+    int nextNullRow = nullIterator.next();
     if (offset.isContiguous()) {
+      final int startOffset = offset.getStartOffset();
+      nullIterator.advanceIfNeeded(startOffset);
       for (int i = 0; i < offset.getCurrentVectorSize(); i++) {
-        retVal[i] = nullValueBitmap.get(i + offset.getStartOffset());
+        final int row = i + startOffset;
+        if (row == nextNullRow) {
+          retVal[i] = true;
+          if (nullIterator.hasNext()) {
+            nextNullRow = nullIterator.next();
+          } else {
+            break;
+          }
+        } else {
+          retVal[i] = false;
+        }
       }
     } else {
+      final int[] currentOffsets = offset.getOffsets();
+      nullIterator.advanceIfNeeded(currentOffsets[0]);
       for (int i = 0; i < offset.getCurrentVectorSize(); i++) {
-        retVal[i] = nullValueBitmap.get(offset.getOffsets()[i]);
+        final int row = currentOffsets[i];
+        if (row == nextNullRow) {
```

Comment:
You were correct on this, I have fixed the implementation to only use `advanceIfNeeded` and `peekNext` which I think makes it a bit easier to follow, though maybe not quite as optimal since it might result in a few additional `advanceIfNeeded` operations. I also added tests for this method that catch this, since it wasn't actually covered otherwise previously.